### DISCUSSION TITLE
Issue with x-show Ignoring x-transition.delay and Displaying Immediately

### DISCUSSION THREAD

**Discussion by desvu** [alpinejs/alpine#4432] [3. Feature ideas]
I noticed that x-show removes display: none as soon as the show variable is set to true, ignoring x-transition.delay. Why not add a delay before displaying the element if a delay is specified? This would at least help resolve issues when two toggled elements share a common container.

**Comment by ekwoka** [comment]
This behavior is how actual css would mostly make it work. You have to have it show otherwise you can't do the transtion.

You can solve the container thing in smarter ways like having them in a single cell grid.

**Reply by desvu** [reply]
I understand that for a transition (I'm talking about CSS transition), the element needs to have a display value different from none. But why show it immediately if there is a delay? Wouldn't it be simpler to wait for the delay before displaying the element and starting the transition?

**Reply by ekwoka** [reply]
Because that's how CSS transitions work.

there is no transition if the thing is not transitioning, and it can't transition from none, so if it's none, the transition never happens, and the delay is part of the transition.

**Reply by desvu** [reply]
It seems we were talking about different things. Alright, let's approach this from a different angle. I found a good existing example that demonstrates the issue:

[UIkit Overlay](https://getuikit.com/docs/search#overlay)

What do you think, is it possible to replicate this with Alpine.js without any hacks?
I couldn't get it to work, and the problem is that x-transition doesn't properly account for delay if they are present. x-show instantly shows the element as soon as the condition is met, without considering the delay.

Here's how I imagine the solution: Alpine.js should detect if there is a delay in x-transition, and if so, use setTimeout to wait for the delay instead of relying on transition-delay. Once the delay time has passed, it can set the element's display property to block, and then perform the transition for a fade-in effect.

**Reply by desvu** [reply]
[codepen](https://codepen.io/alex922/pen/wvVQEGx)

**Reply by SimoTod** [reply]
I think there's some confusion on what x-transition does. It applies css transitions and classes. The delay there is a css transition delay as well, it doesn't delay the execution of the js code.
In general, the transition cares about what happens to the div, not the interactions with what is around.
Js (so Alpine) is used to change the state but the presentation layer is a css concern, it's better not to use js for stuff that is natively supported.

Do you want to divs to be in the same place, just stack them and use the transition to hide show them accordingly (use css, for example position absolute but there may be other way as well, I'm not a frontend developers).

Quick example: https://codepen.io/SimoTod/pen/XWvyyBQ

**Reply by desvu** [reply]
Thank you for the proposed solution, but I was already aware of it in advance. It's a poor option for several reasons:

Absolute positioning forces the element to leave the natural flow.
Simultaneous activation of two transitions looks awful when real UI elements (such as buttons, text, etc.) are involved instead of regular colored blocks. At some point during the transition, they overlap with each other, which looks terrible! Of course, you could use opaque backgrounds, but in my case, that’s more of a half-measure.
It seems I misunderstood the goals of Alpine—it looks like it's a framework for simple things.

**Reply by SimoTod** [reply]
It's indeed more aimed at doing single things. 

About being a poor solution, it's probably an opinion. The timeout solution was poor from my point of view, for example (it's complex, not elegant and it's force you to reinvent what css already does).

As i said, there are probably better solutions than using absolute positioning but I'm not a frontender, the point is that css already gives you the tool to achieve that. 

Also, how you handle the transition is up to you. Usually you want the fade to happen immediately and the show to transition gradually from an aesthetic point of view, otherwise the whole animation could annoy the final user by being too slow. 

But you can delay the css transition as well once the div are stacked.

As a last resource, if you really love that solution, you can create your own directive that just delay the removal of display none (it's not hard if you look at the documentation) and it's likely to have less bug than forcing an existing directive to do something that is not meant to do.

**Reply by desvu** [reply]
Thanks for the help! But it seems that the best solution in my case will be to ditch Alpine. It's not even about the transitions; there are more significant issues that I’ll have to deal with. The first thing that comes to mind is performing some actions after closing a modal, because Alpine doesn't support x-transition hooks. I used Alpine + Tailwind for a while — it was an interesting experience! But now I see that, in the end, I'm reinventing the wheel.

**Reply by SimoTod** [reply]
Yeah, the missing hooks are something many people want. You can hack something with the native browser events but I personally find the transitionend a bit unreliable because Alpine removes the css properties just a tick too early for half of the times so the browser doesn't fire the event. For quick effects, it's probably better to use transitionstart.

For example, https://codepen.io/SimoTod/pen/PoMxVzv does something similar to UIKit but ultimately, you should probably go with what you are more comfortable with and make you more productive, especially if you are the only person working on a project.

**Reply by ekwoka** [reply]
> What do you think, is it possible to replicate this with Alpine.js without any hacks?

Yes.
https://awesomealpine.com/play?share=6ksGkczsDj-5

I'd normally do it like this.

> It seems I misunderstood the goals of Alpine—it looks like it's a framework for simple things.

It's not for "simple things", it just is providing simple pieces, to reduce opinion leakage. It's meant as a simple way to hook up reactivity, not something to dictate the way you do everything.

I'm not had any similar issues with doing thing after a modal closes...but I don't just use x-show with divs to make modals, so it might just be how much simpler using dialog elements is.
